lib/wibox/widget/template.lua

```diff
@@ -0,0 +1,431 @@
+            conn.src_obj:disconnect_signal("property::"..conn.src_prop, conn.apply)
+        end
+    end
+    self._private.connections = {}
+end
+
+--- Give the internal widget instance.
+-- @treturn widget The widget instance.
+-- @method get_widget
+-- @hidden
+function template:get_widget()
+    return lazy_load_child(self)
+end
+
+--- Set the update_callback property.
+-- @tparam function update_callback The new callback function.
+-- @method set_update_callback
+-- @hidden
+function template:set_update_callback(update_callback)
+    assert(type(update_callback) == "function" or update_callback == nil)
+
+    self._private.update_callback = update_callback
+end
+
+--- Hack to allow automatic update of the widget at construction time.
+-- This is supposed to be a setter for an `update_now` property. This property
+-- however doesn't exist. We use this setter in the scope of the widget
+-- construction from wibox.widget internals to offer an easy way for the user to
+-- ask for the widget to be update right after its construction.
+-- Note : The update is not instantly called, but is registered as a normal
+-- update from the `:update()` method.
+-- @tparam[opt] boolean update_now Update the widget now.
+-- @method set_update_now
+-- @hidden
+function template:set_update_now(update_now)
+    if update_now then
+        self:update()
+    end
+end
+
+
+--- Create a new `wibox.widget.template` instance using the same template.
+--
+-- This copy will be blank. Note that `set_property`, `bind_property` or
+-- `update_callback` from `self` will **NOT** be transferred over to the copy.
+--
+-- The following example is a new widget to list a bunch of colors. It uses
+-- `wibox.widget.template` to let the module user define their own color
+-- widget. It does so by cloning the original template into new instances. This
+-- example doesn't handle removing or updating them to keep the size small.
+--
+--@DOC_wibox_widget_template_clone1_EXAMPLE@
+--
+-- @method clone
+-- @treturn wibox.widget.template The copy.
+function template:clone()
+    return template {
+        template = self._private.widget_template
+    }
+end
+
+--- Create a new `wibox.widget.template` instance.
+-- @tparam[opt] table args
+-- @tparam[opt] table|widget args.template The widget template to use.
+-- @tparam[opt] function args.update_callback The callback function to update
+--   the widget.
+-- @tparam[opt] boolean args.update_now Update the widget after its
+--   construction. This will call the `:update()` method with no parameter.
+-- @treturn wibox.widget.template The new instance.
+-- @constructorfct wibox.widget.template
+function template.new(args)
+    args = args or {}
+
+    local ret = wbase.make_widget(nil, nil, { enable_properties = true })
+
+    gtable.crush(ret, template, true)
+    ret._private.connections = {}
+
+    ret:set_template(args.template)
+    ret:set_update_callback(args.update_callback)
+    ret:set_update_now(args.update_now)
```

Comment:
Does the `update_callback` still make sense with the new `:set_property` and `:bind_property` methods?

It seems like these 2 new methods are a more powerful and easier way to achieve what the user would have manually written in the `update_callback`.

Reply:
As far as I'm aware, those two things are not interchangeable.
`:set_property` is for the developer who creates a widget that _uses_ the template, to assign values at named roles in an unknown template.
`update_callback` is for the user who _creates_ the template, to do stuff when values change.

E.g. `awful.widget.tasklist` will call `:set_property` internally, the user that creates a tasklist will assign an `update_callback` to change stuff in the template dynamically.